Connect refactor

hotpocket opened the pull request:
Created `GrpcSerializerBase` and moved as much as the proto types would allow into it from `GrpcSerializer`.

start runme server on port 9999 to test this
`./bin/runme server --address localhost:9999 --runner --tls ~/git/forks/vscode-runme/tls/`

seems to do everything the `GrpcSerializer` does.

hotpocket committed:
move reusable grpc components into a base class

hotpocket committed:
Add ConnectSerializer and the code to select and run it. Refactor kernel slightly to use new base class.

jlewi commented on src/extension/serializer.ts:
```diff
@@ -691,6 +681,60 @@ export class GrpcSerializer extends SerializerBase {
     return Boolean(sessionOutputId)
   }
 
+  // unable to implement marshal methods here, the underlying object structure is the same but
```
What are the data types you are referring to that prevents marshal methods from being implemented here?

hotpocket replied:
I couldn't pass a type via a type parameter, there are no common base classes, union types are a no go, etc ... which means we have type errors when parsing the code of the functions.  The structure may be the same but the data types nested within differ sometimes. A good example of this is in the method `marshalCellOutputs` where in ts_proto we have `UInt32Value` and `string` whereas es_proto we have `number` and `bigint`.  Aside from all of this there was one anomaly I couldn't explain. In `marshalFrontmatter` the return structure has a `tag` property in ts_proto, but in es_proto it's considered invalid to have it there. I wonder if there was some version of the proto that changed between compiles/transpiles?

jlewi replied:
I think this comment is no longer relevant now that we are going with a single class using the es library.

jlewi commented on src/extension/serializer.ts:
```diff
@@ -964,9 +1002,9 @@ export class GrpcSerializer extends SerializerBase {
   }
 
-  private static marshalCellExecutionSummary(
+  protected static marshalCellExecutionSummary(
```
Why is this method being changed from private to protected? Where does it get invoked from?

hotpocket replied:
the plan was to add method signatures in the base to ensure they got implemented in children.  I can switch back to private or add the method signature. LMK.

hotpocket replied:
this method sig has been reverted.

jlewi commented on src/extension/serializer.ts:
```diff
@@ -1009,16 +1047,395 @@ export class GrpcSerializer extends SerializerBase {
+      httpVersion: '2',
+      nodeOptions: {
+        rejectUnauthorized: false, // allow self-signed certificates
+        cert: fs.readFileSync(certPath),
```
What happens if the file doesn't exist?

hotpocket replied:
the extension fails to initialize. I will add some error checking here.

hotpocket replied:
resolving this. can re-comment if this needs additional attention.

jlewi commented on src/extension/serializer.ts:
```diff
@@ -1009,16 +1047,395 @@ export class GrpcSerializer extends SerializerBase {
+    //   (b) the runme binary server default: '~/.config/runme/tls'
+    //   (c) other, manually configured
+
+    // I will assume 2a here
```
How does the gRPC serializer determine the appropriate location for the cert?

hotpocket replied:
(the short version): very close to the same as I do, using the path the extension resides plus a `'/tls'` string.  
(the long version): The GrpcSerializer gets it from the the transport which is constructed in the kernel which gets it from the constructor arg, which gets it from vscode extension base path, which has an '/tls' string added via the `getTLSDir` configuration function. I can't, however, access the kernel methods because they are protected & private. I do have access to the same vscode context, and configuration function, so I could do the following replacement if desirable within the `ConnectSerializer.createSerializerClient` method. 
`const tlsPath = path.normalize(__dirname + '/../tls')`
replace with
`const tlsPath = getTLSDir(Uri.parse(this.context.extensionPath))`
of course this assumes you import `getTLSDir` from the configuration file

hotpocket replied:
Since I've already done the legwork to suss out the aforementioned call, I'll commit it as an update to this PR soon.

jlewi replied:
Should we move transport construction into the kernel? Should we change the kernel API to expose some other information like the tlsDir?

What's the cleanest most robust way to do this?

hotpocket replied:
the way to remove the most code is likely to expose the `getTLS` method in kernelServer.ts so the certs can be used when constructing the client. I don't know if there are security concerns in doing this, it seems very possible. If we are trying to move away from using the file system because we want to run in the browser then we are in new territory and a decision has to be made on where this tls data is to come from.

jlewi commented on src/extension/serializer.ts:
```diff
@@ -1009,16 +1047,395 @@ export class GrpcSerializer extends SerializerBase {
+
+    return createPromiseClient(ParserService, createGrpcTransport(grpcOptions))
+
+    // return createConnectTransport({
```
Delete commented out code.

hotpocket replied:
I am currently creating a grpc transport for testing reasons. I thought we were going to go back to a `createConnectTransport` in the near future, this is the reason I left this dangling.  Should I attempt to revive the connect server and swap out the transports here or leave it as grpc for now and delete this ?

jlewi commented on src/extension/serializer.ts:
```diff
@@ -1009,16 +1047,395 @@ export class GrpcSerializer extends SerializerBase {
+    if (!notebook) {
+      return this.printCell('⚠️ __Error__: no cells found!')
+    }
+    return notebook as any as Serializer.Notebook // ugly cast :(
```
Are you missing logic ala https://github.com/stateful/vscode-runme/blob/127028fb126a1512e0ac4dec84df0ba83db5f566/src/extension/serializer.ts#L296 
To guess the language id of the cells?

hotpocket replied:
I don't believe so.  `deserializeNotebook` calls `reviveNotebook` so lang detection is handled after `reviveNotebook` returns AFAICT.

jlewi replied:
Now that we are just going with a single class and just changing the library can we eliminate most of this diff?

jlewi commented on src/extension/serializer.ts:
```diff
@@ -1009,16 +1047,395 @@ export class GrpcSerializer extends SerializerBase {
+    return notebook as any as Serializer.Notebook // ugly cast :(
+  }
+
+  protected applyIdentity(data: es_proto.Notebook): es_proto.Notebook {
```
For applyIdentity, switchLifecycleIdentity, and other methods that are duplicated in the ConnectSerializer; if you converted the es_proto to a timostamp proto then you could just call  a function that was also used by the gRPCSerializer

e.g.

applyIdentityESProto(data: es_proto.Notebook) {
   ts_proto = convert(data)
   new_ts_proto = applyIdentityTSProto(ts_proto)
   es_proto = convert_ts_to_es(data)

   return es_proto
}

hotpocket replied:
it seems like it would be this way or a refactor to replace ts_proto with es_proto as mentioned in a subsequent comment.  let me know which approach you feel is the best spend of time here.

jlewi replied:
This comment is no longer relevant now that we going with a single class using just es_proto. I wrote this before it occurred to me thats what we should do.

jlewi commented:
It seems like a lot of code duplication right now is the same function written against each version of the proto.

e.g. You have two versions of marshalFrontMatter one which es_proto and one which returns ts_proto. Rather than duplicate the logic, could you have a single function which returns a ts_proto and then just convert the ts_proto to the es_proto if you need an es_proto?

Similarly if you have   somFunc(ts_proto) and you want to create someFunc(es_proto) could you just convert the es_proto to ts_proto and then invoke someFunc(ts_proto)?

Duplicating core business logic (E.g. marshalFrontMatter) will likely lead to problems over time as we will forgot to update both implementations. So the more we can minimize duplication the better.

jlewi commented:
Actually I think we can make this much simpler. We don't need to maintain two separate classes GrpcSerializer and ConnectSerializer in order to continue to support gRPC. We can have a single class and just use either the Grpc or connect transport. Put another way, we can effectively just refactor the GRPCSerializer to use the bufbuild/es library rather than the timostamp library but continue to use gRPC.

Migrating to es is something we want to do independent of using connect protocol because the es library is the newer maintained library (see https://github.com/stateful/runme/issues/641).

hotpocket committed:
handle error on missing tls similar to kernelServer. Add missing call to applyIdentity

hotpocket committed:
add missing config block

hotpocket commented:
> Actually I think we can make this much simpler. We don't need to maintain two separate classes GrpcSerializer and ConnectSerializer in order to continue to support gRPC. We can have a single class and just use either the Grpc or connect transport. Put another way, we can effectively just refactor the GRPCSerializer to use the bufbuild/es library rather than the timostamp library but continue to use gRPC.
> 
> Migrating to es is something we want to do independent of using connect protocol because the es library is the newer maintained library (see [stateful/runme#641](https://github.com/stateful/runme/issues/641)).

So if I'm interpreting this correctly, effectively what you're saying is rename ConnectSerializer to GrpcSerializer, add a method called `setProtocol` or something like that, use this to construct the transport, and call it good?

jlewi commented:
@hotpocket The bulk of the changes will be refactoring the code to use the es library for the protos rather than timostamm library.  The other thing that will need to change is how the client and transport get started. Right now the client and transport gets instantiated in [IServer](https://github.com/stateful/vscode-runme/blob/127028fb126a1512e0ac4dec84df0ba83db5f566/src/extension/server/kernelServer.ts#L64). We can't simply change that code to use the es GRpc Transport because the execute server will still use the timostamp library. While we could also update it to use the es in one go I think its better to try to limit the scope of the change

jlewi commented:
How's this coming?

hotpocket commented:
> How's this coming? 

My apologies, I have not been able to get to this until today. 

I was hoping to get some more detail on what level of disruption/modification of the existing code base is tolerable. The prior method we discussed was to duplicate and modify code to get a usable non-interfering connect serializer(I did some refactoring in kernel.ts for clarity, but this is not strictly necessary).  

If we want to combine things, we run back into integrations issues. If we want a drop in replacement I imagine we would use the server via the constructor if we receive it, otherwise obtain a transport some other way. Depending on the protocol we want to use (connect|grpc)  call something like `setProtocol` which would create the right client & transport, which was what I was thinking when I mentioned this in a prior comment.  If we don't want to duplicate logic, and want to keep our method signatures the same, it seems we would need to construct three proto objects instead of one, per method that uses protos (and some of these are nested and in loops like the marshal methods).  (1) obtain the proto (something like `Notebook.clone(data)` currently), (2) convert to es proto (via code you have shared previously), do some work (as per the current ConnectSerializer), and (3) to convert back to ts for a valid return type. I haven't done any performance testing, and I know premature optimization is the root of all evil, but this seems less than ideal. There doesn't seem to be an ideal solution. The only hard blocker I see in all of this is the missing [tag property](https://github.com/hotpocket/vscode-runme/blob/ea26204d53a052d0e9e06bf2f4cd9b6c3b4ca3a6/src/extension/serializer.ts#L1350) mentioned in a prior comment.  If we must have this property, this needs to be addressed.

Refactoring from ts to es is already done via the ConnectSerializer.  I have already talked about the refactoring concerns, but maybe we can have an instance of a ConnectSerializer inside of each instance of GrpcSerializer and forward calls & convert proto types? Server use and conversion overhead notwithstanding. 

If we want to change if/how a server gets started (maybe depending on config options?) we should discuss this as this falls outside of the scope of implementing a different serialization mechanism I think.

We can of course get on a call to hash this out if that is a better medium to address these issues.

jlewi commented:
The intent was to preserve existing functionality; i.e. the ability to use gRPC. Your current code proves we can do this by using the es library and a gRPC transport.

The limitation of the current PR is that in order to have two code paths (1 for es and 1 for timostamp) you are duplicating a bunch of code. e.g `applyIdentity`.  This will be a burden to maintain.

The way to fix that is to drop support for timostamp in the serializer. Basically, replace GRPCSerializer with your connect serializer. 

To reduce the amount of change in one PR; I would focus this PR on replacing the use of the timostamp library with the es library in the Serializer. That is I would not introduce an option to use the connect transport or set the serializer address in this PR. I would suggest moving that into a follow on PR.

hotpocket committed:
creating the maskedNotebook using `new` causes cells to be copied by reference, causing masked data to be written to the main markdown file which causes data loss

hotpocket committed:
Merge branch 'main' into connect-refactor

hotpocket committed:
getOutputsUri no longer resides on GrpcSerializer and thus broke this test.

hotpocket committed:
replace GrpcSerializer with ConnectSerializer and fix types as needed.

hotpocket committed:
client init in serializer.ts now calls `getTLSEnabled()` which needs`isWindows()`

hotpocket committed:
i have no idea what's going on here but this makes it work.

hotpocket commented:
I have needed to fix some of the tests as a result of the `GrpcSerializerBase` refactor. There very well may be other issues. It may be worth eliminating this base class and integrating everything into the new GrpcSerializer class.  That is, unless we will want to support multiple serializers that talk to grpc services.  In general, testing likely needs some attention after all these changes.

hotpocket committed:
Merge branch 'main' into connect-refactor

hotpocket committed:
fix struct and data items that differ in new implementation.
One test still fails whose solution is unclear "should backfill the output type for buffers". A comment has been added to start the discussion around this test.

hotpocket committed:
fix typo

hotpocket commented on tests/extension/serializer.test.ts:
```diff
@@ -487,6 +487,8 @@ describe('GrpcSerializer', () => {
       const items = cells.outputs[0].items
       expect(items.length).toBe(2)
       items.forEach((item) => {
+        // item.data is a Uint8Array in both the es and ts proto type
```
With the new es implementation `data` is not getting loaded into the object returned from the Notebook constructor.  This seems to be because the data that it receives is not valid (the structure is wrong) per the proto ts bindings.  The weird thing is, this is not valid per the old ts proto bindings either, but the object returned from the ts clone method would happily seed it anyway in the returned object. 

I don't know if this exists for historical reasons, or the fixture is wrong, or some other case outside of my visibility.

jlewi replied:
Is there a bug in [marshallCellOutputs](https://github.com/hotpocket/vscode-runme/blob/af02abc975894a9b7b323b2f1f6f2e57a5adf789/src/extension/serializer.ts#L1035). It looks like that function is responsible for populating the vscode NotebookCellOutputItems from the es proto outputs so seems like that could be the culprit?

That said did you step through the code to see where data stops being set?

hotpocket replied:
**The short version:**

The test seems to be testing something it never should have and the new es type strips out this invalid data so the test must be modified/corrected accordingly

**The long version:**

The data gets lost [here](https://github.com/hotpocket/vscode-runme/blob/af02abc975894a9b7b323b2f1f6f2e57a5adf789/src/extension/serializer.ts#L956) but not [here](https://github.com/hotpocket/vscode-runme/blob/61525b14112e9acb106c2e05033938a6472f8774/src/extension/serializer.ts#L856)

If you look at the es proto in `node_modules/@buf/stateful_runme.bufbuild_es/runme/parser/v1/parser_pb.d.ts` you get 
```
/**
 * @generated from field: bytes data = 1;
 */
data: Uint8Array;
```

Per the proto (and thus both the ts & es bindings) `cells[0].outputs[0].items[0].data` is an array of bytes, not a nested structure that contains an additional data property, which is what [the fixture contains](https://github.com/hotpocket/vscode-runme/blob/61525b14112e9acb106c2e05033938a6472f8774/tests/extension/fixtures/marshalNotebook.json#L30) that is being used as the notebook data.  This seems to be a valid vscode NotebookData structure, but constructing the es Notebook via a constructor, vs the ts type via a static clone method passes through differing data. 
 It seems the timostamm-ts clone will seed data in the returned object that is invalid per the proto but the es constructor will not.  Timostamm Notebook.clone(data) is now notebookInstance.clone() in es, and of course you can't clone an object that doesn't exist, thus the constructor. 

The following images should illustrate this clearly.

Timostamm-ts marshalNotebook process: 
1. the received vs code NotebookData object
2. the cloned Notebook proto
3. the augmented notebook that gets returned
![Screenshot from 2024-12-14 20-20-31](https://github.com/user-attachments/assets/ab9571c7-d32c-4565-bcdf-74b6f65242e1)

Buildbuf_es marshalNotebook process: 
1. the received vs code NotebookData object
2. the cloned Notebook proto
3. the augmented notebook that gets returned
![Screenshot from 2024-12-14 20-16-18](https://github.com/user-attachments/assets/38b018e7-2fc7-4f0d-ad49-d41156512bde)

As you can see the test is looking for `item.data.type`, which is invalid, because data is just an array of bytes.  The es object strips this out, so you have to find the type on the item itself via `item.type` ...


I'll submit a fix for this shortly.

jlewi replied:
So it looks like the code you pointed at where the data is getting lost in the new es code
https://github.com/hotpocket/vscode-runme/blob/af02abc975894a9b7b323b2f1f6f2e57a5adf789/src/extension/serializer.ts#L956
but not in the old ts timeostamp
https://github.com/hotpocket/vscode-runme/blob/61525b14112e9acb106c2e05033938a6472f8774/src/extension/serializer.ts#L856
is where the VSCodeNotebookData is being converted into the Proto Data structure.
In the original code it looks like it is just "casting"  the VSCode Ddata to the Timostamp proto by calling Clone.
https://github.com/hotpocket/vscode-runme/blob/61525b14112e9acb106c2e05033938a6472f8774/src/extension/serializer.ts#L856

As you've noticed in other places the generated code for the ES and Timostamp protos is not the same. So it looks like TS built in casting feature doesn't work for casting the ES data proto. So I think you need to write a proper conversion routine.

Can you write a conversion function and include unittests to verify its working correctly?

hotpocket replied:
It strikes me as incorrect for the test to be trying to access a property of the es proto type object that is not valid. Shouldn't we attempt to keep the es proto object inline with the type definition as per the .proto file (and consequently the typescript bindings)?

The vs-code NotebookData object has `...item.data.type` which is translated/converted to the es-proto Notebook property `...item.type`.  What problem exists that needs additional conversion logic?

see [the update to the test](https://github.com/hotpocket/vscode-runme/pull/4/files#diff-7b0bc79f78f0b7e61b3760261e729614bce3cd51aaf7847266ef165057497d23L484) for full context.

jlewi replied:
> It strikes me as incorrect for the test to be trying to access a property of the es proto type object that is not valid
That's correct. I think I misunderstood what you meant by the data gets lost here but not here.


The TS and ES proto types are not the same.  

Here's the class def for the ES proto

```
export declare class CellOutputItem extends Message<CellOutputItem> {
  /**
   * @generated from field: bytes data = 1;
   */
  data: Uint8Array;

  /**
   * @generated from field: string type = 2;
   */
  type: string;

  /**
   * @generated from field: string mime = 3;
   */
  mime: string;

  constructor(data?: PartialMessage<CellOutputItem>);

  static readonly runtime: typeof proto3;
  static readonly typeName = "runme.parser.v1.CellOutputItem";
  static readonly fields: FieldList;

  static fromBinary(bytes: Uint8Array, options?: Partial<BinaryReadOptions>): CellOutputItem;

  static fromJson(jsonValue: JsonValue, options?: Partial<JsonReadOptions>): CellOutputItem;

  static fromJsonString(jsonString: string, options?: Partial<JsonReadOptions>): CellOutputItem;

  static equals(a: CellOutputItem | PlainMessage<CellOutputItem> | undefined, b: CellOutputItem | PlainMessage<CellOutputItem> | undefined): boolean;
}
```

And for the TS proto

```
/**
 * @generated from protobuf message runme.parser.v1.CellOutputItem
 */
export interface CellOutputItem {
    /**
     * @generated from protobuf field: bytes data = 1;
     */
    data: Uint8Array;
    /**
     * @generated from protobuf field: string type = 2;
     */
    type: string;
    /**
     * @generated from protobuf field: string mime = 3;
     */
    mime: string;
}
```

That said in both cases  there isn't a property type isn't nested on.

What happens if you ran the test on main; does it pass? 

> This seems to be a valid vscode NotebookData structure, but constructing the es Notebook via a constructor, vs the ts type via a static clone method passes through differing data.
It seems the timostamm-ts clone will seed data in the returned object that is invalid per the proto but the es constructor will not.

I think you are right. So IUC it looks like the test isn't using a typed object for VSCodeNotebookData. Its just passing around raw JSON and coariscing that into VSCodeData? So is there a bug in the testData
https://github.com/hotpocket/vscode-runme/blob/61525b14112e9acb106c2e05033938a6472f8774/tests/extension/fixtures/marshalNotebook.json#L29 ? And its just that with the TS protos the buggy data gets passed around?

If your pretty confident its a bug in the test; then I would suggest breaking it out into a separate fix and submitting it to the upstream repository separately. This will make it much easier to review and also ensure the updated test passes on the old code.

hotpocket replied:
>What happens if you ran the test on main; does it pass?

It passes, because the timostam-ts type accepts structure that is not valid and then the test tests for that structure. The structure (`item.data.type`) is part of the vs-code NotebookData type but not part of the proto Notebook type, which instead has the structure (`item.type`).  So it works because the proto type and test are both bad/invalid in a specific way (it seeds and looks for structure that is invalid). The es proto type rejects the bad data, and thus the test fails. 

> So IUC it looks like the test isn't using a typed object for VSCodeNotebookData. Its just passing around raw JSON and coariscing that into VSCodeData? So is there a bug in the testData

The test data AFAICT is a valid structure for a vscode/NotebookData object, and mirrors what happens in a normal situation. It's not a genuine instance of this class but I doubt that matters. The es_proto Notebook constructor actually wants a `PartialMessage<Notebook>` but it gets a `vscode/NotebookData`, which is only allowed because of the cast in marshalNotebook `new Notebook(data as any)`.  Putting aside the force cast(not my code), the resulting proto_es Notebook object constructed __does__ get a valid Buffer object created as the value of `item[x].data` (this is some sort of proto bindings dark magic) which is tested in `marshalCellOutputs` via `item.type = item.data.buffer ? 'Buffer' : typeof item.data` and seeds the value the test __should__ be looking for.  

>And its just that with the TS protos the buggy data gets passed around?

It's not buggy data, it's valid data from a different type (see above). 

I will create a separate branch & PR to address this bug in the serializer test.

hotpocket replied:
PR submitted.

hotpocket replied:
for the curious:

the `type` property of `item.data` seems to be an artifact of serializing the `NotebookData` object via `JSON.stringify()` ... but when `JSON.parse` is called it doesn't recreate the `Buffer` object, instead it creates a plan js Object. The PR ( https://github.com/stateful/vscode-runme/pull/1875 ) for this clearly illustrates this last ah ha moment for this thread.

hotpocket replied:
this PR has been merged. closing this conversation thread.

jlewi commented on src/extension/serializer.ts:
```diff
@@ -473,36 +473,23 @@ export class WasmSerializer extends SerializerBase {
-  private client?: ParserServiceClient
-  protected ready: ReadyPromise
+// no common ancestor, any type used for protos
+export abstract class GrpcSerializerBase extends SerializerBase {
```
As you suggested in the comment we should get rid of the GrpcSerializerBase class as we aren't using it anymore. 

You should try to make this as easy as possible for the runme folks to review but try to minimize the number of changes so they can focus their attention on the most substantitive changes.

jlewi commented:
Seems like the remaining work is

1. Get the tests to pass
2. Minimize the diff by getting rid of the unnecessary new base class and other refactorings that are no longer needed now that we have a single class
3. Figure out the best solution for initializing the transport and in particular getting the suitable TLS information

hotpocket commented:
> Seems like the remaining work is
> 
> 1. Get the tests to pass
> 2. Minimize the diff by getting rid of the unnecessary new base class and other refactorings that are no longer needed now that we have a single class
> 3. Figure out the best solution for initializing the transport and in particular getting the suitable TLS information

1 & 2 are "easy". 3 is an executive decision. How do we get a cert in a secure way if we're not reading it from the file system?  I'm using the method as gleaned from the server process, as per previous discussions. What do you want this process to look like?  see  comment earlier in this PR where tls is discussed.

hotpocket committed:
fix invalid proto prop reference

hotpocket committed:
combine grpc base and grpc serializer. reorder changes to minimize PR diff.

hotpocket committed:
more diff consolidation for PR

hotpocket commented on src/extension/serializer.ts:
```diff
@@ -991,18 +1035,33 @@ export class GrpcSerializer extends SerializerBase {
+    })
+    let res
+    try {
+      res = await this.client.deserialize(deserializeRequest)
```
I added extra error handling here because without it there was no error shown in vscode when the the request for deserialization failed. the old grpc serializer didn't do any additional error checking ... i can remove it if we feel safe to do so

hotpocket commented on src/extension/messages/platformRequest/saveCellExecution.ts:
```diff
@@ -187,7 +187,7 @@ export default async function saveCellExecution(
         },
       }
-      const result = await graphClient.mutate(mutation)
+      const result = await graphClient.mutate(mutation as MutationOptions)
```
I am concerned about this because I have no idea why this is needed.  The typescript error on this was very hairy.  If someone smarter than me knows what is going on here please let me know.

jlewi replied:
Why would this code need to change at all if its running on head? What was the error you got?

hotpocket replied:
```Argument of type '{ mutation: TypedDocumentNode<CreateExtensionCellOutputMutation, Exact<{ input: ReporterInput; }>>; variables: { ...; }; }' is not assignable to parameter of type 'MutationOptions<CreateExtensionCellOutputMutation, Exact<{ input: ReporterInput; }>, DefaultContext, ApolloCache<any>>'.
  The types of 'variables.input.notebook.cells' are incompatible between these types.
    Type '{ outputs: { items: CellOutputItem[]; metadata: { [key: string]: string; }; processInfo?: CellOutputProcessInfo | undefined; }[]; kind: CellKind; ... 4 more ...; executionSummary?: CellExecutionSummary | undefined; }[]' is not assignable to type 'ReporterCellInput[]'.
      Type '{ outputs: { items: CellOutputItem[]; metadata: { [key: string]: string; }; processInfo?: CellOutputProcessInfo; }[]; kind: CellKind; ... 4 more ...; executionSummary?: CellExecutionSummary; }' is not assignable to type 'ReporterCellInput'.
        Types of property 'executionSummary' are incompatible.
          Type 'CellExecutionSummary | undefined' is not assignable to type 'InputMaybe<ReporterCellExecutionSummaryInput> | undefined'.
            Type 'CellExecutionSummary' is not assignable to type 'ReporterCellExecutionSummaryInput'.
              Types of property 'executionOrder' are incompatible.
                Type 'number | undefined' is not assignable to type 'InputMaybe<UInt32> | undefined'.
                  Type 'number' has no properties in common with type 'UInt32'.ts(2345)```

hotpocket replied:
Ok, after testing the wall thoroughly with my forehead I have found a loose brick. It seems this graphql stuff was generated and it seems from the timostam-ts proto type. On line 2014 in graphql.ts there exists a line `executionOrder?: InputMaybe<UInt32>;` which conforms to the timostam-ts type Notebook.Cell.CellExecutionSummary.executionOrder which is `executionOrder?: UInt32Value;` which is in contrast to the es type, which is, as you'd guess, a `number` ... so my guess is this graphql.ts needs to be regenerated.

Can you say more about what you mean by "running on head"?  I am on this branch looking for issues caused by this code.  I assume you mean the head of the upstream repo? If this is not correct please elaborate.

jlewi replied:
Yes I mean making sure your branch is up to date with upstream changes

jlewi replied:
Can this graphql code continue to use the timeostamp protos in this PR?

hotpocket replied:
> Can this graphql code continue to use the timeostamp protos in this PR?

This brings us back to the "can we make these two coexist" question as I see it, which was attempted for quite a long time until it was determined they could not coexist, and the original GrpcSerializer was modified so that there would not be two implementations. From what I can tell this will need to be regenerated, or we can force cast (as I have done in this branch) and pray it doesn't cause any runtime errors.  I ran the tests for this and didn't get any failures, but I'm not sure this specific case is being tested. 

I don't see where graphql.ts gets generated.  Should I post a question on discord?

jlewi replied:
Oh sorry I didn't realize from the diff this was using grpcSerializer. Yeah if its using grpcSerializer it might need to change. Sure ask in discord.

hotpocket commented:
i feel we're getting close here....

hotpocket committed:
pr diff cleanup

jlewi commented on src/extension/serializer.ts:
```diff
@@ -886,9 +930,9 @@ export class GrpcSerializer extends SerializerBase {
+      return new Frontmatter({
         category: '',
-        tag: '',
+        // tag: '',  // buf-es does not have tag property, but the old timostam/prototype-ts did
```
I think you may need to update the dependency for the stateful es library.
It looks like tag was added in September
https://github.com/stateful/runme/commit/77ace522e46b2fb111e9800bf5adb8369f7f648b

I don't see a direct dependency on that library in package.json
https://github.com/hotpocket/vscode-runme/blob/f9f3e1bb4be8c146bf8b15e20afd4667943b7e05/package.json#L1365

So I wonder if its getting pulled in indirectly via jlewi. 

At any rate try running

```
npm install @buf/stateful_runme.bufbuild_es@1.10.0-20241021161352-11c142f88aee.1
```

To update the dependency and see if that version has the tag field.

hotpocket replied:
this initially seemed to work with the exception of a conflict with the Notebook type that is defined in your package, because your type doesn't have the tag property presumably.  I tried to resolve this but it seemed to cause dependency issues. my guess is because your package depends on a certain version but this version is different than that. This spidered into other connect and es dependencies as I tried to fix things. Could you update the version in your package and I could then update your package version here?

here is the typescript error I got

```
Argument of type 'import("/home/brandon/git/forks/vscode-runme/node_modules/@buf/jlewi_foyle.bufbuild_es/node_modules/@buf/stateful_runme.bufbuild_es/runme/parser/v1/parser_pb").Notebook' is not assignable to parameter of type 'import("/home/brandon/git/forks/vscode-runme/node_modules/@buf/stateful_runme.bufbuild_es/runme/parser/v1/parser_pb").Notebook'.
  Types of property 'frontmatter' are incompatible.
    Type 'import("/home/brandon/git/forks/vscode-runme/node_modules/@buf/jlewi_foyle.bufbuild_es/node_modules/@buf/stateful_runme.bufbuild_es/runme/parser/v1/parser_pb").Frontmatter | undefined' is not assignable to type 'import("/home/brandon/git/forks/vscode-runme/node_modules/@buf/stateful_runme.bufbuild_es/runme/parser/v1/parser_pb").Frontmatter | undefined'.
      Property 'tag' is missing in type 'import("/home/brandon/git/forks/vscode-runme/node_modules/@buf/jlewi_foyle.bufbuild_es/node_modules/@buf/stateful_runme.bufbuild_es/runme/parser/v1/parser_pb").Frontmatter' but required in type 'import("/home/brandon/git/forks/vscode-runme/node_modules/@buf/stateful_runme.bufbuild_es/runme/parser/v1/parser_pb").Frontmatter'.ts(2345)
parser_pb.d.ts(488, 3): 'tag' is declared here.
```

jlewi replied:
Done.
https://buf.build/jlewi/foyle/commits

Per https://buf.build/jlewi/foyle/deps seems to be depending on latest version of stateful now.

hotpocket replied:
this seems to resolve the issue. Thanks!

jlewi commented on src/extension/serializer.ts:
```diff
@@ -512,8 +521,44 @@ export class GrpcSerializer extends SerializerBase {
+      try {
+        if (!getTLSEnabled()) {
+          return {}
+        }
```
How about adding a method to KernelServer which returns the pem `async getPem()` and then just call that function here so the logic for getting the PEM is written in a single place inside Kernel?

hotpocket replied:
These two methods in kernelServer.ts would need to be made public `private static async getTLS(tlsDir: string) `, and `protected getTLSDir(): string`. Adding other method(s) would cause duplication.  Are we ok w/ making this public?  This has been discussed in [this](https://github.com/hotpocket/vscode-runme/pull/4#discussion_r1872619773) thread above.  Thoughts?

jlewi replied:
Is there a reason to not make it public? If it avoids code duplications what's the downside?

hotpocket replied:
It just makes me feel a bit dirty to enable the server to serve up it's certs to anyone that asks for them.  I'll make these methods public and make a note of it in the final PR to have it reviewed.

hotpocket replied:
this is done in the most recent commit.  let me know if there are any other issues you see here.

jlewi replied:
If you don't want to expose the methods. Then the other option is to create a second routine on the KernelServer that can return a connect es transport (in addition) to the gRPC transport. I thought that might lead to more duplicate code but maybe that's cleaner like you say than exposing TLSDir, TLS, and serverUrl.

hotpocket committed:
Merge branch 'main' into connect-refactor

hotpocket committed:
fix old schema via foyle npm update

hotpocket committed:
Merge branch 'main' into connect-refactor

hotpocket committed:
expose TLS methods to avoid code duplication

jlewi commented on src/extension/serializer.ts:
```diff
@@ -886,9 +925,9 @@ export class GrpcSerializer extends SerializerBase {
+      return new Frontmatter({
         category: '',
-        tag: '',
+        tag: '', // buf-es does not have tag property, but the old timostam/prototype-ts did
```
Is the comment still valid?

hotpocket replied:
fixed

jlewi commented on src/extension/serializer.ts:
```diff
@@ -917,19 +956,19 @@ export class GrpcSerializer extends SerializerBase {
       },
       category: '',
-      tag: '',
+      tag: '', // es_proto does not have a tag property, but timostamm-ts did
```
Is the comment still valid?

hotpocket replied:
woops :) .  will remove that

jlewi commented:
Nice work. I think this is almost there.

1. I'd suggest going ahead and opening a PR with the updates to serializertest.ts in the upstream repo
2. I think there are some minor edits about the comment that need to be fixed
3. I think after fixing 2 you can open up a PR against the main repo and see what Sebastien thinks about the best way to instantiate the transport

hotpocket committed:
remove stale comment

hotpocket commented:
> 1. I'd suggest going ahead and opening a PR with the updates to serializertest.ts in the upstream repo

This is already in the PR queue - https://github.com/stateful/vscode-runme/pull/1875

> 2. I think there are some minor edits about the comment that need to be fixed

fixed

> 3. I think after fixing 2 you can open up a PR against the main repo and see what Sebastien thinks about the best way to instantiate the transport

Submitted here https://github.com/stateful/vscode-runme/pull/1876

jlewi commented:
Fantastic. Looks like the first one got merged.
